### PR TITLE
Typechain Polkadot Grant 2

### PR DESCRIPTION
**Pull request by TtomaS7**
### Project Abstract
[Preview](https://github.com/Supercolony-net/Open-Grants-Program/blob/feature/typechain-follow-up/applications/typechain-polkadot-follow-up.md)

[Previous Typechain W3F grant](https://github.com/w3f/Grants-Program/pull/992)

We attempt to bring type safety to working with Polkadot contracts at the front-end side of development. Based on contracts ABIs' content, we generate corresponding runtime code in addition to TypeScript types definitions for the contracts, so that developers deal with pre-typed methods on the contract.

Generating such code automatically with CLI brings convenience to the dApps development, as well as speeds up required codebase updates, when contracts are changed in pre-release stage.

We have decided to add a new tool to the 3rd milestone that will help developers to easily integrate typechain to their projects, and it will also be compatible with Redspot (if redspot was used for compiling projects, you can just leave most of the things).

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying?

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $50,000, 3 approvals
- [x] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] I have read and understood the [FAQs](https://github.com/w3f/Grants-Program/blob/master/docs/faq.md), [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/grant_guidelines_per_category.md) and [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/announcement-guidelines.md).
- [x] A BTC, Ethereum (USDT/USDC/DAI) or Polkadot/Kusama (aUSD) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledge the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).
- [x] I prefer the discussion of this application to be in a private Element/Matrix channel. My username is: ``@toma.sadova:matrix.org``

### How Did You Hear About our grants program?

- [x] Other: Previous successful applications for `OpenBrush` and `Sol2Ink`

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Just a quick reminder here to sign the updated version of our Terms & Conditions. I will reach out via elements in case I have any other questions.

**Review by Noc2 — APPROVED** [review]
Thanks for the update. I’m happy to go ahead with it. And sorry for the delay, I thought I had already approved it.

**Comment by cruikshankss** [comment]
Hi @TtomaS7, this proposal is really interesting. I am excited to use this myself. However, could you double-check the total costs, which should be 97,300 by my calculations (instead of 97,000)?

**Comment by TtomaS7** [comment]
Thank you for the feedback!

oh yeah, it’s my bad...I have changed it already

**Review by cruikshankss — DISMISSED** [review]

File: applications/typechain-polkadot-follow-up.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+* **Costs:** 52,500 USD
+
+| Number | Deliverable | Specification |
+|--------| ----------- | ------------- |
+| 0a     | License | MIT |
+| 0b     | Documentation | We will cover new-added features in documentation and usage examples. |
+| 1      | Precise methods definitions | Refine definitions and bahavior of contracts methods (i.e. methods' arguments and returns), depending on namespace, call options and properties of the method, like `payable` & `mutable`. E.g. preamptive querying for transaction calls, controlled by a call options flag. |
+| 2      | Methods' names | Format methods' names in the output from original `MethodTrait::method_name` to more user-friendly `methodName` naming scheme, while resolving overlap in formatted names. |
+| 3      | Contract classes extension | Extend generated contract classes with useful properties, normally available on the contract (e.g. address, name, signer, etc.). Rely on usage experience in doing so. |
+| 4      | IDE hints | Prepare generated code to have more informattive IDE hints, based on both JSDoc and output typesystem itself (if needed). Rely on usage experience in doing so. |
```
Comment (takahser) [inline applications/typechain-polkadot-follow-up.md:203]:
```suggestion
| 4      | IDE hints | Prepare generated code to have more informative IDE hints, based on both JSDoc and the output typesystem itself (if needed). Rely on user experience in doing so. |
```

File: applications/typechain-polkadot-follow-up.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+| 5      | NPM Package | Translate package's code to JavaScript upon deployment. Provide informattive CLI, when needed. Make sure to have a cross-platform CLI support. |
```
Comment (takahser) [inline applications/typechain-polkadot-follow-up.md:204]:
```suggestion
| 5      | NPM Package | Translate package's code to JavaScript upon deployment. Provide informative CLI, when needed. Make sure to have cross-platform CLI support. |
```

File: applications/typechain-polkadot-follow-up.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+* **Estimated duration:** 8 weeks
+* **FTE:**  2
+* **Costs:** 44,800 USD
+
+| Number | Deliverable | Specification |
+| ------ | ----------- | ------------- |
+| 0a | License | MIT |
+| 0b | Documentation | We will cover new-added features in documentation and usage examples. |
+| 1 | Investigation | Broaden types definitions for methods arguments and return values (to full coverage). |
+| 2 | Parser & generators modules | Design and implement a new parser module for ABI JSON to work with different versions of the ABI. Parser's output structure serves as an input for generators. Refactor, replace inlne generation with parser to generator flow. |
```
Comment (takahser) [inline applications/typechain-polkadot-follow-up.md:185]:
```suggestion
| 2 | Parser & generators modules | Design and implement a new parser module for ABI JSON to work with different versions of the ABI. Parser's output structure serves as an input for generators. Refactor, replace inline generation with parser to generator flow. |
```

File: applications/typechain-polkadot-follow-up.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+There is a whole side to every project in the ecosystem - that is the front-end side of contract integration, - that will find use for Typechain-Polkadot during development process.
+
+- Who is your target audience?
+
+Our main target audience is front-end and smart-contract developers of Polkadot projects.
+
+- What need(s) does your project meet?
+
+Our project speeds up the front-end integration of Polkadot/Substrate/Kusama contracts. As well as brings type safety to developers' code, when dealing with contracts. For smart contract developers, it simplifies integration test writing.
```
Comment (takahser) [inline applications/typechain-polkadot-follow-up.md:66]:
```suggestion
Our project speeds up the front-end integration of Polkadot/Substrate/Kusama contracts. It also brings type safety to developers' code, when dealing with contracts. For smart contract developers, it simplifies integration test writing.
```

File: applications/typechain-polkadot-follow-up.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+```bash
+    npx typechain-compiler --config config.json
+```
+
+Also you can set some additional arguments like `--noCompile`, `--noTypechain`, `--release`
+
+Config interface will be something like this:
+```typescript
+export interface Config {
+    projectFiles: string[]; // Path to all project files, everystring in glob format
```
Comment (takahser) [inline applications/typechain-polkadot-follow-up.md:38]: What do you mean by "glob format"?

**Review by takahser — CHANGES_REQUESTED** [review]
@TtomaS7 thanks for the submission of this grant application.

I have a couple of questions:

- Is there any specific reason the daily rate is at 600.- for M3 while it is 560.- for M1 and M2?
- At M2, "3 | Contract deployment", does this mean it will be possible to generate the types directly from the `*.contract` file, without the need of an ABI? If so, what's the motivation behind that, since both are build artifacts, so usually there will always be an ABI if there is a `*.contract` file.
- At M3, "1 Precise methods definitions", what do you mean by that? 
- At M3, "2 Methods' names", why don't you implement it like that directly in M2 already?
- At M3, "3 Contract classes extension", is there a reason not to include these "useful properties, normally available on the contract" already in M2?
- At M3, "4 IDE hints", what do you mean by "output typesystem"? Also, does JSDoc not just mean that you have useful inline comments to describe the types and methods, which is something you should do anyway in M2 already?
- At M3, "5 NPM Package", I assume this includes the publishing of the npm package, is that correct? If so, could you update that?
- At M3, "6 Typechain-compiler", what's the difference to the CLI provided in M2?

In general, I'm mostly fine with M2, while I think M3 looks bloated with deliveries that rather describe non-functional requirements that IMHO should be part of M2 already.

File: applications/typechain-polkadot-follow-up.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+### Ecosystem Fit
+
+- Where and how does your project fit into the ecosystem?
+
+There is a whole side to every project in the ecosystem - that is the front-end side of contract integration, - that will find use for Typechain-Polkadot during development process.
+
+- Who is your target audience?
+
+Our main target audience is front-end and smart-contract developers of Polkadot projects.
```
Comment (takahser) [inline applications/typechain-polkadot-follow-up.md:62]:
```suggestion
Our main target audience is consisting of front-end and smart-contract developers of Polkadot projects.
```

**Review by takahser — CHANGES_REQUESTED** [review]

File: applications/typechain-polkadot-follow-up.md (hunk 5 above)
Comment (varex83) [reply applications/typechain-polkadot-follow-up.md:38]: "Globs" are the patterns you type when you do stuff like ls *.js on the command line, or put build/* in a .gitignore file

**Review by Noc2 — APPROVED** [review]

**Comment by TtomaS7** [comment]
> * Is there any specific reason the daily rate is at 600.- for M3 while it is 560.- for M1 and M2?

Yeah, M3 has FTE:2.5 and M2 has FTE: 2
The rate of the developers are different (who will work on it), this is why I increase the daily rate

**Comment by varex83** [comment]
Thanks for your questions. We have reviewed it and made changes to our grant proposal. So check them out, please.
The reason for splitting this part up into 2 parts was to do the main work in the second milestone and then improve existing things in Milestone 3. But after our suggestion, we decided to reconstruct the Milestones.
 At M2, “3 | Contract deployment”, does this mean it will be possible to generate the types directly from the *.contract file, without the need of an ABI? If so, what’s the motivation behind that, since both are build artifacts, so usually there will always be an ABI if there is a *.contract file.
We have clarified in our grant proposal that we will add availability to deploy contracts with the Constructors field, using *.contract files.
At M3, “1 Precise methods definitions”, what do you mean by that?
At M3, “2 Methods’ names”, why don’t you implement it like that directly in M2 already?
At M3, “3 Contract classes extension”, is there a reason not to include these “useful properties, normally available on the contract” already in M2?
At M3, “4 IDE hints”, what do you mean by “output typesystem”? Also, does JSDoc not just mean that you have useful inline comments to describe the types and methods, which is something you should do anyway in M2 already?
We reviewed it and decided to make it part of the 2nd milestone, so all these features aim to create a better user experience. About the Contract classes extension, we have updated the description in our grant proposal. It’s about making valuable fields and methods like withSigner or withAddress .
At M3, “5 NPM Package”, I assume this includes the publishing of the npm package, is that correct? If so, could you update that?
We are now using the GitHub registry to publish it, but we will publish it on npmjs to improve the user experience.
At M3, “6 Typechain-compiler”, what’s the difference to the CLI provided in M2?
We already have CLI since 1st milestone, but Typechain only generates code to interact with the contract, and it’s not very useful for big projects. Typechain-compiler is a tool that solves this problem: it can compile all contracts in the project, bring all ABIs and .contract files in one place and then run typechain. In the future, we will add more features to it, like a CLI wrapper to run user scripts. If you need more details regarding our next steps on Typechain, please have a look on Future plans section
Thank you for your detailed feedback and questions, it gets us more ideas on how we could improve our Milestones. I appreciate it!

**Review by cruikshankss — DISMISSED** [review]
Hi @TtomaS7, thanks for answering those great Qs from @takahser. I understand this TypeChain followup proposal now much better after completing your OpenBrush M6 evaluation. Firstly, the ABI JSON parser module seems like it would be very fun to build! :) Also, I would appreciate some clarity on how your tool relates to https://paritytech.github.io/contracts-ui/. 

As a frontend React/Next.js dev, I know from experience it's quite complicated to understand how data comes in from the blockchain to the frontend, and may I suggest that you make some educational materials with the keywords such as "Web3 Tech Stack", "Full-stack Polkadot/Substrate dApp tutorial", "Full front-end development tutorial with Polkadot and Substrate for beginners", "As a frontend dev, how can I get data from Polkadot/Substrate blockchains to the frontend?". I actually found this current job I have because I was searching for "web3 tech stack" on google. Before that, I had watched a lecture from Next.js Conf about Next.js 12 and there was [a lecture from Nader Dabit on web3 tech stacks](https://www.youtube.com/watch?v=f9XRH7bjV8M)! (also see his ["complete guide"](https://dev.to/edge-and-node/the-complete-guide-to-full-stack-web3-development-4g74) for more info) Since then, I find it's still quite complicated to onboard frontend devs & teams to the ecosystem.

Also, may I make some suggestions? The npm packaging system is quite horribly broken, but slightly less broken if you use `yarn` instead of `npx` for versioning/config/packages if that's possible for you. Facebook/Meta made/maintains [yarn](https://yarnpkg.com/), for some background (and of course Facebook also created React). The creator of Node.js event loop (Ryan Dahl), even has departed from the npm ecosystem and has created the [Deno](https://deno.land/) ecosystem (which is intentionally incompatible with npm due to its brokenness). Deno is secure, type-safe and written in Rust. If it's possible, I think it would be very cool to see a frontend project in this ecosystem utilizing Deno (or at least on the road to eventually using it)!

By the way, I see plugins are in your future plans and I personally would like to advocate for a VSCode TypeChain plugin because I've found this IDE has the best Rust plugins. XD 🦀

**Review by Noc2 — APPROVED** [review]

**Comment by TtomaS7** [comment]
@cruikshankss Thank you so much for your review and suggestion! we will keep it on track.

**Comment by TtomaS7** [comment]
@takahser do you need any details from our side, after answering questions?

**Comment by cruikshankss** [comment]
@TtomaS7 I just re-shared your application with the committee. Thanks for your patience.

**Review by takahser — CHANGES_REQUESTED** [review]
@varex83 @TtomaS7 
Thanks for the update and the clarifications and sorry for the delay here.

To me, it seems like you added 4a, 4b and 7 to M2 and updating the price from 44.8k$ to 77k$. However, imho these deliveries do not justify a price increase of >30k$. Let me explain:

- 4a Contract classes extension: tbh I'm still confused about this one in general. I know you [pointed out](https://github.com/w3f/Grants-Program/pull/1105#issuecomment-1213316163) that

    > It’s about making valuable fields and methods like withSigner or withAddress

    but I'm not sure what you mean by that.

- 4b Methods' names: here you're just switching from `snake_case` to `camelCase`. While I totally support your efforts to use `camelCase` here, since it abides to the general TypeScript coding conventions, I don't understand why you wouldn't already include this qualitative requirement into the original implementation. So why implement it in `snake_case` first, only to convert it to `camelCase` after?
- 7 Branding: We don't usually support design work as part of a grant. Also, what exactly do you want to improve in the README? In general, I think it's good practice to update the README through development.

For these reasons, I'm fine with supporting M1 & M2 at their original price points but not at the current levels.

**Comment by varex83** [comment]
Thanks, @takahser, for your response. 

- About 4a Contract classes extension is to make additional fields that will help developers interact with a contract; for example, field `withSigner` could change contract' signer without defining a new contract instance, so it will decrease code re-usage and make interacting with contracts easier from multi-user perspective, and also will help test contracts in the better way 
- About 4b Methods' names, we decided to add 4a, 4b, and so on as sub-task because they are not such big as needed to be full tasks because they aren't time-consuming, so we moved them from the 3rd milestone to 2nd. 
-  About 7 README - is about adding a new logotype, gifs, and a better description of newly added features to improve the user experience. It's time-consuming and needs the involvement of our design unit. You can see how we maintain documentation in OpenBrush.

Generally speaking, soon, there will be the first commercial project that uses Typechain. [FUNDRs](https://www.fundrs.app/) from AllianceBlock, which already got success on Avalanche, will use Typechain and sol2ink to transpile Solidity smartcontracts' code to ink! 

We are happy to hear your thoughts and speed up the process of grant acceptance!
I really appreciate any help you can provide.

**Comment by TtomaS7** [comment]
Hello, W3F team!
Let’s define how we could move forward with the grant

**Comment by Noc2** [comment]
Thanks for chasing us. I pinged the rest of the committee again.

**Comment by TtomaS7** [comment]
thank you! I appreciate it

**Comment by TtomaS7** [comment]
hello!
@Noc2 do you have any updates?

**Review by takahser — CHANGES_REQUESTED** [review]
@TtomaS7 and @varex83 thanks for sharing your thoughts. However, I think you still haven't taken any stance against my previous [comment](https://github.com/w3f/Grants-Program/pull/1105#pullrequestreview-1081839819):

> To me, it seems like you added 4a, 4b and 7 to M2 and updating the price from 44.8k$ to 77k$. However, imho these deliveries do not justify a price increase of >30k$. 

Hence, my [position is still](https://github.com/w3f/Grants-Program/pull/1105#pullrequestreview-1081839819):

> For these reasons, I'm fine with supporting M1 & M2 at their original price points but not at the current levels.

**Review by takahser — APPROVED** [review]
Thanks for the update and the reduction of the price. Although I'm aware of the fact that there is still a slight premium to the original M2 price ($52.2k instead of $44.8k), I'm happy to go along with it.

**Review by laboon — APPROVED** [review]

**Comment by TtomaS7** [comment]
Hello, W3F team! 
Do you have any updates?

**Review by Noc2 — APPROVED** [review]

**Review by alxs — APPROVED** [review]

**Pull request merged**